File: packages/yaacli/yaacli/desktop/protocol.py
"""Versioned wire protocol shared by YAACLI Desktop and its sidecar."""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

PROTOCOL_VERSION = 1
MAX_MESSAGE_BYTES = 1024 * 1024
MAX_PENDING_REQUESTS = 64
MAX_ATTACHMENTS = 32

Identifier = Annotated[str, StringConstraints(min_length=1, max_length=128)]


class ProtocolModel(BaseModel):
    """Strict base class for protocol values."""

    model_config = ConfigDict(extra="forbid")


class InputPartType(StrEnum):
    TEXT = "text"
    FILE = "file"
    IMAGE = "image"


class RunStatus(StrEnum):
    IDLE = "idle"
    RUNNING = "running"
    WAITING_APPROVAL = "waiting_approval"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    FAILED = "failed"


class InputPart(ProtocolModel):
    type: InputPartType
    text: str | None = None
    path: str | None = None
    media_type: str | None = None
    name: str | None = None
    data_base64: str | None = None


class WorkspaceInfo(ProtocolModel):
    id: Identifier
    path: str
    name: str
    available: bool = True
    guidance_sources: list[str] = Field(default_factory=list)
    config_sources: list[str] = Field(default_factory=list)
    git_branch: str | None = None


class SessionSummary(ProtocolModel):
    id: Identifier
    name: str
    latest_user_prompt: str = ""
    updated_at: str = ""
    workspace_id: Identifier
    model: str = ""
    archived: bool = False


class SessionSnapshot(ProtocolModel):
    session: SessionSummary
    transcript: list[dict[str, Any]] = Field(default_factory=list)
    run_status: RunStatus = RunStatus.IDLE


class RunInfo(ProtocolModel):
    id: Identifier
    session_id: Identifier
    workspace_id: Identifier
    status: RunStatus
    steerable: bool = False


class ApprovalDecision(StrEnum):
    APPROVE_ONCE = "approve_once"
    APPROVE_SESSION = "approve_session"
    DENY = "deny"


class ApprovalRequest(ProtocolModel):
    id: Identifier
    workspace_id: Identifier
    session_id: Identifier
    run_id: Identifier
    tool_call_id: Identifier
    tool_name: str
    summary: str
    risk: str = "runtime_review"
    decisions: list[ApprovalDecision] = Field(
        default_factory=lambda: [
            ApprovalDecision.APPROVE_ONCE,
            ApprovalDecision.APPROVE_SESSION,
            ApprovalDecision.DENY,
        ]
    )


class FileChange(ProtocolModel):
    path: str
    change_type: Literal["added", "modified", "deleted", "renamed"]
    old_path: str | None = None
    diff: str | None = None
    diff_available: bool = False
    binary: bool = False


class UsageInfo(ProtocolModel):
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    context_window: int | None = None
    cost: float | None = None


class RuntimeHealth(ProtocolModel):
    status: Literal["starting", "ready", "unavailable", "stopping"]
    workspace_id: str | None = None
    active_run_id: str | None = None
    message: str | None = None


class ProtocolCapabilities(ProtocolModel):
    commands: list[str]
    events: list[str]
    max_message_bytes: int = MAX_MESSAGE_BYTES
    max_attachments: int = MAX_ATTACHMENTS
    steering: bool = True
    approvals: bool = True


class HandshakeEnvelope(ProtocolModel):
    protocol_version: Literal[1] = PROTOCOL_VERSION
    type: Literal["handshake"] = "handshake"
    runtime_version: str
    capabilities: ProtocolCapabilities


class RequestEnvelope(ProtocolModel):
    protocol_version: Literal[1] = PROTOCOL_VERSION
    type: Literal["request"] = "request"
    request_id: Identifier
    command: Identifier
    payload: dict[str, Any] = Field(default_factory=dict)


class ErrorInfo(ProtocolModel):
    code: str
    message: str
    retryable: bool = False


class ResponseEnvelope(ProtocolModel):
    protocol_version: Literal[1] = PROTOCOL_VERSION
    type: Literal["response"] = "response"
    request_id: Identifier
    ok: bool
    payload: dict[str, Any] | None = None
    error: ErrorInfo | None = None


class EventEnvelope(ProtocolModel):
    protocol_version: Literal[1] = PROTOCOL_VERSION
    type: Literal["event"] = "event"
    event: Identifier
    payload: dict[str, Any] = Field(default_factory=dict)
    workspace_id: str | None = None
    session_id: str | None = None
    run_id: str | None = None
    sequence: int | None = Field(default=None, ge=0)


WireEnvelope = HandshakeEnvelope | RequestEnvelope | ResponseEnvelope | EventEnvelope


SUPPORTED_COMMANDS = [
    "runtime.health",
    "workspace.open",
    "session.create",
    "session.list",
    "session.list_archived",
    "session.load",
    "session.rename",
    "session.archive",
    "session.restore",
    "run.start",
    "run.cancel",
    "run.steer",
    "approval.resolve",
    "config.get",
    "config.update",
    "runtime.shutdown",
]

SUPPORTED_EVENTS = [
    "runtime.health",
    "run.started",
    "run.phase",
    "text.delta",
    "thinking.delta",
    "tool.started",
    "tool.completed",
    "task.updated",
    "usage.updated",
    "file.changed",
    "approval.requested",
    "approval.resolved",
    "steering.acknowledged",
    "run.completed",
    "run.cancelled",
    "run.failed",
]

File: apps/yaacli-desktop/src/protocol.ts
import { z } from "zod";

export const PROTOCOL_VERSION = 1 as const;

const identifier = z.string().min(1).max(128);

export const inputPartSchema = z
  .object({
    type: z.enum(["text", "file", "image"]),
    text: z.string().nullable().optional(),
    path: z.string().nullable().optional(),
    media_type: z.string().nullable().optional(),
    name: z.string().nullable().optional(),
    data_base64: z.string().nullable().optional(),
  })
  .strict();

export type InputPart = z.infer<typeof inputPartSchema>;

export const workspaceInfoSchema = z
  .object({
    id: identifier,
    path: z.string(),
    name: z.string(),
    available: z.boolean().default(true),
    guidance_sources: z.array(z.string()).default([]),
    config_sources: z.array(z.string()).default([]),
    git_branch: z.string().nullable().optional(),
  })
  .strict();

export const sessionSummarySchema = z
  .object({
    id: identifier,
    name: z.string(),
    latest_user_prompt: z.string().default(""),
    updated_at: z.string().default(""),
    workspace_id: identifier,
    model: z.string().default(""),
    archived: z.boolean().default(false),
  })
  .strict();

export const runStatusSchema = z.enum([
  "idle",
  "running",
  "waiting_approval",
  "completed",
  "cancelled",
  "failed",
]);

export const sessionSnapshotSchema = z
  .object({
    session: sessionSummarySchema,
    transcript: z.array(z.record(z.string(), z.unknown())).default([]),
    run_status: runStatusSchema.default("idle"),
  })
  .strict();

export const approvalDecisionSchema = z.enum([
  "approve_once",
  "approve_session",
  "deny",
]);

export const approvalRequestSchema = z
  .object({
    id: identifier,
    workspace_id: identifier,
    session_id: identifier,
    run_id: identifier,
    tool_call_id: identifier,
    tool_name: z.string(),
    summary: z.string(),
    risk: z.string(),
    decisions: z.array(approvalDecisionSchema),
  })
  .strict();

export const fileChangeSchema = z
  .object({
    path: z.string(),
    change_type: z.enum(["added", "modified", "deleted", "renamed"]),
    old_path: z.string().nullable().optional(),
    diff: z.string().nullable().optional(),
    diff_available: z.boolean().default(false),
    binary: z.boolean().default(false),
  })
  .strict();

export const usageInfoSchema = z
  .object({
    input_tokens: z.number().int().nonnegative().default(0),
    output_tokens: z.number().int().nonnegative().default(0),
    total_tokens: z.number().int().nonnegative().default(0),
    context_window: z.number().int().positive().nullable().optional(),
    cost: z.number().nonnegative().nullable().optional(),
  })
  .strict();

export const protocolCapabilitiesSchema = z
  .object({
    commands: z.array(z.string()),
    events: z.array(z.string()),
    max_message_bytes: z.number().int().positive(),
    max_attachments: z.number().int().positive(),
    steering: z.boolean(),
    approvals: z.boolean(),
  })
  .strict();

export const handshakeEnvelopeSchema = z
  .object({
    protocol_version: z.literal(PROTOCOL_VERSION),
    type: z.literal("handshake"),
    runtime_version: z.string(),
    capabilities: protocolCapabilitiesSchema,
  })
  .strict();

export const requestEnvelopeSchema = z
  .object({
    protocol_version: z.literal(PROTOCOL_VERSION),
    type: z.literal("request"),
    request_id: identifier,
    command: identifier,
    payload: z.record(z.string(), z.unknown()),
  })
  .strict();

export const errorInfoSchema = z
  .object({
    code: z.string(),
    message: z.string(),
    retryable: z.boolean().default(false),
  })
  .strict();

export const responseEnvelopeSchema = z
  .object({
    protocol_version: z.literal(PROTOCOL_VERSION),
    type: z.literal("response"),
    request_id: identifier,
    ok: z.boolean(),
    payload: z.record(z.string(), z.unknown()).nullable().optional(),
    error: errorInfoSchema.nullable().optional(),
  })
  .strict();

export const eventEnvelopeSchema = z
  .object({
    protocol_version: z.literal(PROTOCOL_VERSION),
    type: z.literal("event"),
    event: identifier,
    payload: z.record(z.string(), z.unknown()),
    workspace_id: z.string().nullable().optional(),
    session_id: z.string().nullable().optional(),
    run_id: z.string().nullable().optional(),
    sequence: z.number().int().nonnegative().nullable().optional(),
  })
  .strict();

export const wireEnvelopeSchema = z.discriminatedUnion("type", [
  handshakeEnvelopeSchema,
  requestEnvelopeSchema,
  responseEnvelopeSchema,
  eventEnvelopeSchema,
]);

export type WorkspaceInfo = z.infer<typeof workspaceInfoSchema>;
export type SessionSummary = z.infer<typeof sessionSummarySchema>;
export type SessionSnapshot = z.infer<typeof sessionSnapshotSchema>;
export type ApprovalRequest = z.infer<typeof approvalRequestSchema>;
export type FileChange = z.infer<typeof fileChangeSchema>;
export type UsageInfo = z.infer<typeof usageInfoSchema>;
export type RequestEnvelope = z.infer<typeof requestEnvelopeSchema>;
export type ResponseEnvelope = z.infer<typeof responseEnvelopeSchema>;
export type EventEnvelope = z.infer<typeof eventEnvelopeSchema>;
export type WireEnvelope = z.infer<typeof wireEnvelopeSchema>;
